lib/ecto/repo/queryable.ex

```diff
@@ -128,7 +128,14 @@ defmodule Ecto.Repo.Queryable do
 
   defp attach_prefix(query, opts) do
     case Keyword.fetch(opts, :prefix) do
-      {:ok, prefix} -> %{query | prefix: prefix}
+      {:ok, prefix} ->
+        combinations =
+          Enum.map(query.combinations, fn {type, combination_query} ->
+            {type, %{combination_query | prefix: prefix }}
+          end)
+
+        %{query | prefix: prefix, combinations: combinations}
```

Comment:
If the combination itself already has a prefix though, we should most likely respect it. Also, this needs to be done recursively. The combination itself may have other combinations. I think the best place to apply this is inside Ecto.Query.Planner.

Reply:
@josevalim thanks for the suggestion! I've moved logic of prefix attachment for combinations  into a query planner and I've added more tests to cover, for example, the recursive case you mentioned.